Libraries/LibWeb/DOM/Element.cpp

```diff
@@ -1874,6 +1874,54 @@ void Element::invalidate_style_after_attribute_change(FlyString const& attribute
     invalidate_style(StyleInvalidationReason::ElementAttributeChange);
 }
 
+bool Element::is_hidden() const
+{
+    if (layout_node() == nullptr)
+        return true;
+    if ((layout_node()->computed_values().visibility() == CSS::Visibility::Hidden || layout_node()->computed_values().visibility() == CSS::Visibility::Collapse || layout_node()->computed_values().content_visibility() == CSS::ContentVisibility::Hidden))
+        return true;
+    for (ParentNode const* self_or_ancestor = this; self_or_ancestor; self_or_ancestor = self_or_ancestor->parent_or_shadow_host()) {
+        if (self_or_ancestor->is_element() && static_cast<DOM::Element const*>(self_or_ancestor)->aria_hidden() == "true")
+            return true;
+    }
+    return false;
+}
+
+bool Element::has_hidden_ancestor() const
+{
+    for (ParentNode const* self_or_ancestor = this; self_or_ancestor; self_or_ancestor = self_or_ancestor->parent_or_shadow_host()) {
+        if (self_or_ancestor->is_element() && static_cast<DOM::Element const*>(self_or_ancestor)->is_hidden())
+            return true;
+    }
+    return false;
+}
+
+bool Element::is_referenced() const
+{
+    bool is_referenced = false;
+    if (id().has_value()) {
+        this->root().for_each_in_inclusive_subtree_of_type<HTML::HTMLElement>([&](auto& element) {
```

Comment:
Not sure if this is important, but this loop will also iterate over this element. Should an element referencing itself behave the same as a different element referencing it?

Reply:
> Not sure if this is important, but this loop will also iterate over this element. Should an element referencing itself behave the same as a different element referencing it?

I don’t now recall intentionally choosing `for_each_in_inclusive_subtree_of_type` rather than `for_each_in_subtree_of_type` and I don’t believe the spec requirements necessitate that, and changing it doesn’t cause any tests to fail — so I’ve gone ahead and changed it to using `for_each_in_subtree_of_type`.